### PR TITLE
Support failed image reloading on Reader posts

### LINKED ISSUE #12517: PNG images displayed as a grey box in reader posts on bad connectivity
### Expected behavior
Post media will show up or show an indication that they were not loaded. 

### Actual behavior
PNG images show up as grey boxes.

### Steps to reproduce the behavior

1. Enable link conditioning with "Very bad network"
2. Go to Reader and open a post with PNG images.
3. Scroll through the post and see a spinner instead of PNG images.
4. Some of them will be updated to a grey box. 

##### Tested on [device], iOS [version], WPiOS [version]
iPhone 7, iOS: 12.4.1 WPiOS: 13.2.1

IMG_4927

Corresponding Android issue: https://github.com/wordpress-mobile/WordPress-Android/issues/10506 

* Hat tip to the report by @m"

### PR DESCRIPTION
**Pull request by guarani**
To improve the user experience on the Reader screen when images fail to download, [issue 12517 was created](https://github.com/wordpress-mobile/WordPress-iOS/issues/12517) to display a retry button allowing the user to retry the download. Currently, if the image is downloaded successfully, this area is occupied by the image. If the image fails to download for any reason, a blank gray view occupies this area. The ask here is to replace this view with a tappable retry button.

### Demo

| **Before**  | **After** |
| ------------- | ------------- |
| original | demo |

### Testing 

1. Log in with valid credentials and tap the Reader tab
2. Navigate back to the Reader home screen and tap the Search option
3. Search for the following post: https://alexdanco.com/
4. If testing on a physical device, activate the 100% Loss option in the Network link Conditioner (located in Settings)
5. If testing on a simulator, download and activate the Apple Network Link Conditioner to the 100% Loss profile
6. Tap a post that contains images (e.g. "Why I don’t love Light Rail transit" - https://alexdanco.com/2019/09/12/why-i-dont-love-light-rail-transit/)
7. Observe that when the timeout completes, the retry view is now shown in the place of an image that failed to download
8. If an internet connection is reenabled, images should download, as before
9. If an image is tapped, it should either be displayed as a preview or open an in-app browser, as before

### Known Issues

1. SVG images are not supported: https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-541345870
2. Issues covered here: https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-542001552

### Pending

- [ ] Also see the summary of pending issues in https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-544165492
- [ ] Test Reader comments for regressions [1] (see below)
- [ ] Handle cases where image area is too small to fit the retry view
- [x] Check retry view font size
- [x] Add unit tests

[1] - This PR modifies `WPRichTextImage` which is also used on the reader comments screen, so there is a potential for errors to be introduced here. I have been unable to test so far due to not having a blog with comment attachments enabled readily available.

### CODE REVIEW

File: WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift [hunk 1]
```diff
@@ -276,6 +276,15 @@ extension UIColor {
         return .neutral(.shade30)
     }
 
+    /// For now, just for the retry view background
+    static var retryBackground: UIColor {
```
Comment (guarani) [inline WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift]: I added this color from https://github.com/wordpress-mobile/WordPress-iOS/issues/12517#issuecomment-532935835 - is there a reference name for this color?

File: WordPress/Classes/Utility/Media/ImageLoader.swift [hunk 2]
```diff
@@ -61,6 +61,14 @@ import MobileCoreServices
...
     }
 
+    func startLoadingAnimation() {
+        loadingIndicator.startAnimating()
+    }
+
+    func stopLoadingAnimation() {
+        loadingIndicator.stopAnimating()
+    }
+
```
Comment (guarani) [inline WordPress/Classes/Utility/Media/ImageLoader.swift]: Will add access control here.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+class ReaderRetryFailedImageView: UIView {
+
+    private class LinkTextView: UITextView {
+        override func selectionRects(for range: UITextRange) -> [UITextSelectionRect] {
+            return []
+        }
+
+        override func caretRect(for position: UITextPosition) -> CGRect {
+            CGRect.zero.offsetBy(dx: .greatestFiniteMagnitude, dy: .greatestFiniteMagnitude)
```
Comment (guarani) [inline WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Either both methods should omit `return` statements here, or both should use them. My feeling is that it's best here to use `return` statements because that way when a new line is added there is just a single line change in the diff.

File: WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift (hunk 1 above)
Comment (guarani) [reply WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift]: Hi @frosty, just wanted to ask for your input here. The [design](https://github.com/wordpress-mobile/WordPress-iOS/issues/12517#issuecomment-532935835) calls for a "Gray-10 / systemGray6 (iOS 13)" color background for the retry view, and I see we don't have that combination in code. I temporarily added it here, but wondering if you could guide me on how to handle this. I'm also thinking these colors already have semantic names defined somewhere but I'm not sure where to find them.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 3 above)
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Resolved using explicit return statements.

File: WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift (hunk 1 above)
Comment (frosty) [reply WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift]: Hey, thanks for the ping! I think it's okay to add this here – as you say, it doesn't match our existing colors so this looks fine 👍 Thanks for checking. Could we maybe clarify the name + comment slightly by changing it to `imageViewRetryBackground` or something similar?

File: WordPress/Classes/Utility/Media/ImageLoader.swift [hunk 4]
```diff
@@ -242,7 +250,7 @@ import MobileCoreServices
             }
 
             if self.imageView.shouldShowLoadingIndicator {
-                self.loadingIndicator.state = .error
+                self.stopLoadingAnimation()
```
Comment (shiki) [inline WordPress/Classes/Utility/Media/ImageLoader.swift]: ⚠️ I'm not quite sure of this change. Would you mind explaining why this has to be done? 

I think it's causing a regression in parts where `CachedAnimatedImageView` is used. Notice the gray background color is not being shown when using this branch:

`develop` | This branch
--------|-------
IMG_0389        |       IMG_0390

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    override func caretRect(for position: UITextPosition) -> CGRect {
+        return CGRect.zero.offsetBy(dx: .greatestFiniteMagnitude, dy: .greatestFiniteMagnitude)
+    }
+}
+
+class ReaderRetryFailedImageView: UIView {
+
+    @IBOutlet weak private var imageView: UIImageView! {
+        didSet {
```
Comment (shiki) [inline WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: I haven't seen this being done before. Nice idea 👍

**Comment by mattmiklic** [comment]
> @mattmiklic I'd like to request your input on the behavior that was implemented here.
> It's working nicely. However, I am concerned with the touch area of the _Retry_ button. It's pretty small:
> I wonder if we should consider allowing users to tap on everything in that box instead? Like this:

Yes, I think this is a good idea. The blue and underlined text is a helpful hint that the user can tap this area, but making the entire box tappable is the way to go.

Visually it's looking pretty good, but could you double-check the colors that are being used? The background color in particular seems to be too dark. For reference here's what I'd specified in the original issue:

> Colors: `Gray-10` / `systemGray6` (iOS 13) for the background, `textSubtle` for the text and icon, `primary` for the link

Aside, there seems to be a lot of padding above/below those images -- is that the default? I realize it could just be the post being previewed.

File: WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift (hunk 1 above)
Comment (guarani) [reply WordPress/Classes/Extensions/Colors and Styles/UIColor+MurielColors.swift]: Thanks for the reply, @frosty! Will update the name and comment.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    override func selectionRects(for range: UITextRange) -> [UITextSelectionRect] {
+        return []
+    }
+
+    override func caretRect(for position: UITextPosition) -> CGRect {
+        return CGRect.zero.offsetBy(dx: .greatestFiniteMagnitude, dy: .greatestFiniteMagnitude)
+    }
+}
+
+class ReaderRetryFailedImageView: UIView {
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: I agree this is lacking explanation. Fortunately, I'll be removing this class because we're no longer going to need the link itself to be tappable now that the retry view itself will be the touch target.
For completeness' sake though - this was to disable selection on the text view. Allowing tappable links necessitated `isSelectable` to be `true`, which has the undesirable result here of allowing text selection. Casting the cursor caret into the far reaches of the coordinate space didn't appeal to me, but was the most reliable hack after lots of testing.

File: WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift [hunk 7]
```diff
@@ -4,10 +4,17 @@ open class WPRichTextImage: UIControl, WPRichTextMediaAttachment {
...
 
+    private var reloadImage: (() -> Void)?
     var contentURL: URL?
     var linkURL: URL?
 
     @objc fileprivate(set) var imageView: CachedAnimatedImageView
+    private lazy var retryImageView: ReaderRetryFailedImageView = {
+        guard let retryView = Bundle.main.loadNibNamed("ReaderRetryFailedImageView", owner: nil, options: nil)?.first as? ReaderRetryFailedImageView else {
+            fatalError("ReaderRetryFailedImageView xib must exist. This is an error.")
+        }
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift]: Yes - we could use a factory-like static method to do the nib loading.

File: WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift [hunk 8]
```diff
@@ -88,13 +95,29 @@ open class WPRichTextImage: UIControl, WPRichTextMediaAttachment {
...
-        let errorHandler: ((Error?) -> Void)? = { error in
+        let errorHandler: ((Error?) -> Void)? = { [weak self] error in
+            guard let self = self else { return }
+
+            self.retryImageView.delegate = self
+            self.addSubview(self.retryImageView)
+            self.retryImageView.leftAnchor.constraint(equalTo: self.layoutMarginsGuide.leftAnchor).isActive = true
+            self.retryImageView.topAnchor.constraint(equalTo: self.layoutMarginsGuide.topAnchor).isActive = true
+            self.retryImageView.rightAnchor.constraint(equalTo: self.layoutMarginsGuide.rightAnchor).isActive = true
+            self.retryImageView.bottomAnchor.constraint(equalTo: self.layoutMarginsGuide.bottomAnchor).isActive = true
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift]: I'm not sure I follow, could you clarify please? 
`WPRichTextImage` doesn't use an xib but if you're referring to `ReaderRetryFailedImageView`'s xib, I don't see how it would add these constraints either.

File: WordPress/Classes/Utility/Media/ImageLoader.swift [hunk 9]
```diff
@@ -61,6 +61,14 @@ import MobileCoreServices
         imageView.addLoadingIndicator(loadingIndicator, style: .fullView)
     }
 
+    internal func startLoadingAnimation() {
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/Utility/Media/ImageLoader.swift]: Thanks, that's true.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 10]
```diff
@@ -0,0 +1,78 @@
+//
+//  ReaderRetryFailedImageView.swift
+//  WordPress
+//
+//  Created by Paul Von Schrottky on 10/5/19.
+//  Copyright © 2019 WordPress. All rights reserved.
+//
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Thanks, miss that!

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: > This means that we created instances of this class for every image even if they're not needed. Would you please let me know why it behaves like this?

Thanks for catching this - it's because I accessed the lazy property in:
`self.retryImageView.removeFromSuperview()`
I'll update this to ensure it's handled properly.

> Also, the post has 3 images only but there were 6 recorded lines in the log. I didn't investigate further why that is.
I traced this back to:
https://github.com/wordpress-mobile/WordPress-iOS/blob/a9d3de84faae974e8508be6f995bd7eaad8af6d5/WordPress/Classes/ViewRelated/Views/WPRichText/WPTextAttachmentManager.swift#L110
which is called multiple times. This is the same in `develop`, let me know if you'd like me to dig further.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+    @IBOutlet weak private var textView: LinkTextView! {
+        didSet {
+            textView.delegate = self
+            textView.textDragInteraction?.isEnabled = false
+            textView.adjustsFontForContentSizeCategory = true
+        }
+    }
+
+    private let attributedString: NSAttributedString = {
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: I sometimes come across an issue in GitHub where comments I write disappear when a PR is open in multiple tabs - so will recap here.
---
I went for a property because it seemed to succinctly encapsulate the creation of the attributed string. It's also a reference to an object that the `textView` also references, so memory shouldn't be an issue. Either way, I'll update to a function as I don't have a strong opinion on this.

Regarding the naming, since it's the only attributed string here, I felt the enclosing type name provided enough context.

**Comment by guarani** [comment]
> I also found this bug where it never stops trying to load the image:
Interesting find @shiki . I noticed the image that isn't downloading is an SVG image, so I tested this on the App Store version and it's also stuck on the image loading state (eventually the spinner disappears leaving the grey background).
Do you think this is a separate bug? If SVG is not supported, I'm not sure if we have an existing UI somewhere that should be shown for unsupported image formats.

**Comment by guarani** [comment]
By the way, I saw some mention of flaky tests in Slack yesterday and the UI tests are not passing. The error seems unrelated to my code:
> Failed to synthesize event: Failed to scroll to visible (by AX action) Button, identifier: 'Continue Button', label: 'Continue', error: Error kAXErrorCannotComplete performing AXAction 2003 on element AX element pid: 837, elementOrHash.elementID: 140477726346784.8525

I think I saw someone say that a fix would be made to rectify the tests, so that should be solved soon.

**Comment by guarani** [comment]
> Visually it's looking pretty good, but could you double-check the colors that are being used? The background color in particular seems to be too dark. For reference here's what I'd specified in the original issue:

Thanks @mattmiklic. 
You're right that the color looks off, I'm using the `gray10` that was predefined in the code using this palette (sorry it's a bit small):

colors
The color in the design looks similar to `gray0` in this list, could it be that?

---

> Aside, there seems to be a lot of padding above/below those images -- is that the default? I realize it could just be the post being previewed.

I think the post mentioned above may have extra space (I couldn't locate it to verify though). Below is how other posts look:
post_margins

File: WordPress/Classes/Utility/Media/ImageLoader.swift (hunk 2 above)
Comment (guarani) [reply WordPress/Classes/Utility/Media/ImageLoader.swift]: Done

File: WordPress/Classes/Utility/Media/ImageLoader.swift (hunk 4 above)
Comment (guarani) [reply WordPress/Classes/Utility/Media/ImageLoader.swift]: My reading of the code is that the `error` state is used in conjunction with the `errorView`. Since we're not using the error view, and I'd noticed that the loading indicator was not removed from the hierarchy when stopped, I thought it should probably be removed when the retry view is shown.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: I missed the `static` part when reading this last night. There is nothing in the message specific to a particular instance so that could work here. I'll update this later today.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: I believe you're talking about `LinkTextView`. I meant for `ReaderRetryFailedImageView`. 🙂

[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: > I'll update this to ensure it's handled properly.

Thanks! This is good enough for me. As for the `enumerateAttachments()`, let's leave that alone.

File: WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift]: It's more of a question, really. I wasn't sure if it's possible. 🙂

**Comment by shiki** [comment]
> Do you think this is a separate bug? If SVG is not supported, I'm not sure if we have an existing UI somewhere that should be shown for unsupported image formats.

I didn't realize I added an SVG. 😄 I'd consider it as a separate bug. Would you mind creating an issue for it? 

> UI tests are not passing

Thanks for checking. Let's worry about those when we can say that this is ready to be merged to `develop`.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Updated this to a `class` func because a `ReaderRetryFailedImageView` subclass (let's say for a embedded video) could want to override the text to change it to something different.

**Comment by mattmiklic** [comment]
> The color in the design looks similar to `gray0` in this list, could it be that?

Ahh yes that's my bad. It should be gray0 for iOS 12 and below; systemGray6 for iOS 13.

Could you double-check the left/right margins for me? It should be the same as the margins on the paragraphs above and below, but in that latest screenshot the image not found area seems to be more inset from the left/right edges than the text above and below it.

**Comment by guarani** [comment]
Hey @mattmiklic, no worries. 

I'll double check that as part of my effort to give the image not found area a 2:1 aspect ratio, which will handle really tall and thin images, wide and short images, or other irregular sizes.

**Comment by guarani** [comment]
> I didn't realize I added an SVG. 😄 I'd consider it as a separate bug. Would you mind creating an issue for it?
@shiki, I found an existing issue that tracks this: https://github.com/wordpress-mobile/WordPress-iOS/issues/9650

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Thanks! I added a short but descriptive comment.

**Comment by guarani** [comment]
@shiki, @mattmiklic - after looking into both the margins https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-541113848 as well as the possibility of using a fixed size retry view to ensure even tiny images fit their retry button, I've come to the conclusion that changes here would require a significant level of effort.

The logic below would likely need to be changed to handle two cases; when an image has failed and when an image has downloaded successfully.
https://github.com/wordpress-mobile/WordPress-iOS/blob/d52c8d2e4364c6e86c599af81e0fc490d064de80/WordPress/Classes/ViewRelated/Views/WPRichText/WPTextAttachment.swift#L84

The current version of the app does not seem to handle really small image attachments particularly well either, as the activity indicator appears to be sized according to the image's size.

**Comment by guarani** [comment]
Testing the retry button revealed two potential issues.

| Scenario | Expected Result | Actual Result | Comments | Test | App Store |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| [Pinterest button](https://bigtinyworld.com/2019/08/15/okonjima-nature-reserve-the-african-cat-experience/) (bottom of post) | Pinterest button is displayed | Retry view is displayed instead | App Store version displays blank area |  |  |
| [Tap failed images](https://mydearkitcheninhelsinki.com/2018/10/30/phyllo-pie-with-fennel-mushroom-rice-leeks-and-so-many-more-things/) (test w/o internet) | Tapping retry reloads image | In-app browser is displayed instead | Tapping blank area also opens in-app browser | n/a | n/a |

iPhone Xs Max (simulator), iOS 12.2

---

### Scenario N° 1
The Pinterest button issue is not confined to this branch - it exists in the current WordPress App Store version. It is important to note however, that the retry view exacerbates this issue by showing the _retry button_ where the Pinterest button should be. On the App Store version, the issue is much less noticeable because a blank area is shown instead.

### Scenario N° 2
The in-app browser issue is due to a gesture recognizer that handles link taps, but also captures taps on images that fail to download - on both the App Store as well as on this branch. Again, the issue is not too apparent on the App Store version but here it defeats the purpose of the retry button. The gesture recognizer in question is found here:
https://github.com/wordpress-mobile/WordPress-iOS/blob/4cb9f7bd233aa34dd0794a504bb3715ed7f564e6/WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift#L162

### Action Items

In view of the above comments, I think it is important to evaluate whether the user experience of the retry button is not significantly hindered by the above two issues. If so, perhaps we should address them first before implementing the retry view. The above https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-541485310 should also be considered as a possible prerequisite here.

Could I please get your thoughts here @mattmiklic and @shiki?

**Comment by mattmiklic** [comment]
> I agree with you that the _retry view_ exacerbates this problem. We'd probably not want to release this with this uncovered issue.

Agreed about this.

File: WordPress/WordPressTest/ReaderRetryFailedImageViewTests.swift [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import XCTest
+import UIKit
+@testable import WordPress
+
+class ReaderRetryFailedImageViewTests: XCTestCase {
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/WordPressTest/ReaderRetryFailedImageViewTests.swift]: Thanks!

File: WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift [hunk 13]
```diff
@@ -4,10 +4,12 @@ open class WPRichTextImage: UIControl, WPRichTextMediaAttachment {
 
     // MARK: Properties
 
+    private var reloadImage: (() -> Void)?
     var contentURL: URL?
     var linkURL: URL?
 
     @objc fileprivate(set) var imageView: CachedAnimatedImageView
+    private lazy var retryImageView = ReaderRetryFailedImageView.loadFromNib()
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift]: I try to only unit test "public" interfaces so I didn't change `ReaderRetryFailedImageView` to `internal` just for testing purposes. I think views can sometimes be tricky because they can be `private` while still forming part of the public interface to the system. This is where I think UI tests can really fill in the gaps to provide good overall coverage.

File: WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift [hunk 14]
```diff
@@ -0,0 +1,74 @@
+import UIKit
+import Gridicons
+
+/// Responding to retry actions.
+protocol ReaderRetryFailedImageDelegate: AnyObject {
+    func didTapRetry()
+}
+
+/// A view that represents a failed image download and includes a retry action.
+class ReaderRetryFailedImageView: UIControl {
```
[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: Thanks for the video, @shiki. I tested this again before reverting the change and found noticed something strange. 
1. `WPRichTextImage.swift:109`
2. `ImageLoader.swift:129`
3. `ImageLoader.swift:88`
4. `ImageLoader.swift:164`
5. `ImageLoader.swift:190`
6. `ImageLoader.swift:221` - execution doesn't step into the `downloadImage(from:` function 
Execution doesn't reach `UIImageView+Networking.swift:35` - it was on line 93 where I would create a `URLSessionConfiguration` to reduce the network timeout to a couple of seconds for testing purposes, but this code is not getting hit. Not sure if it's Xcode playing up but will try a project clean and retry tonight.
Without this all adding up, although I'm pushing some changes here I'm not requesting your review just yet.

[the remark this answers is not in the document]
Comment (guarani) [reply WordPress/Classes/ViewRelated/Reader/ReaderRetryFailedImageView.swift]: With Scenarios 1 and 3 out of scope, Scenario 2 remains. I'll further clarify this scenario now that I've dug a little deeper.

When an image is tapped in the Reader, the image should always be previewed full-screen, if I understand correctly. However, some images when tapped open in an in-app browser. I tested this in the current App Store build and found the same behavior.

Concerning this PR, I reverted (and pushed) the changes in https://github.com/wordpress-mobile/WordPress-iOS/commit/f585e2f88758292c88b3b85baebe219d58189c3c, but the behavior was the same. In light of the fact that this is present in the App Store version, I feel Scenario 2 is also out of scope. Do you agree, @shiki? 

**App Store v13.3**

**Comment by guarani** [comment]
### Current State

TL;DR; Multiple (existing) issues with how the iOS app renders posts (and their attachments) were uncovered. It was decided with @shiki that they should be investigated before continuing with this PR.

In https://github.com/wordpress-mobile/WordPress-iOS/pull/12634#issuecomment-542001552 above, three scenarios were discussed that affect this PR:
 
- Scenario 1: likely bug - create an issue if needed
- Scenario 2: likely bug - create an issue if needed (note: this could be related to https://git.io/Je4Ns)
- Scenario 3: likely (minor) bug - create an issue if needed

cc @mattmiklic

**Comment by shiki** [comment]
Closing the PR for now. We are going to have improvements for Reader soon and this might be affected. We can always reopen it if we want to in the future.
